File: 03-rucksack-reorganization/python/main.py
def get_item_priority(item: str) -> int:
    return ord(item) - 38 if item.isupper() else ord(item) - 96


def get_rucksacks() -> list[tuple[str]]:
    with open("../input.txt", "r") as f:
        return [
            (line[: len(line) // 2], line[len(line) // 2 :].strip("\n")) for line in f
        ]


def get_common_item(rucksack: tuple[str]) -> str:
    first_comp_set, second_comp_set = set(rucksack[0]), set(rucksack[1])
    try:
        return list(first_comp_set.intersection(second_comp_set))[0]
    except IndexError:
        print(rucksack)


def main():
    total_item_priority = 0
    rucksacks = get_rucksacks()
    for rucksack in rucksacks:
        common_item = get_common_item(rucksack)
        total_item_priority += get_item_priority(common_item)
    print(total_item_priority)


main()

File: 03-rucksack-reorganization/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
	"unicode"

	"github.com/juliangruber/go-intersect/v2"
)

func getItemPriority(item []rune) int {
	if unicode.IsUpper(item[0]) {
		return int(item[0]) - 38
	}
	return int(item[0]) - 96
}

func getRucksacks() [][]string {
	f, err := os.Open("../input.txt")
	if err != nil {
		log.Fatal(err)
	}

	defer f.Close()

	scanner := bufio.NewScanner(f)

	rucksack := [][]string{}
	compartments := []string{}

	for scanner.Scan() {
		cleanLine := strings.Trim(scanner.Text(), "\n")
		halfRucksackLength := len(cleanLine) / 2
		firstCompartment := cleanLine[:halfRucksackLength]
		secondCompartment := cleanLine[halfRucksackLength:]
		compartments = append(compartments, firstCompartment, secondCompartment)
		rucksack = append(rucksack, compartments)
		compartments = []string{}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	return rucksack
}

func getCommonItem(rucksack []string) string {
	commonItem := intersect.SimpleGeneric[string](strings.Split(rucksack[0], ""), strings.Split(rucksack[1], ""))
	return commonItem[0]
}

func main() {
	totalItemPriority := 0
	rucksacks := getRucksacks()
	for _, rucksack := range rucksacks {
		commonItem := getCommonItem(rucksack)
		totalItemPriority += getItemPriority([]rune(commonItem))
	}
	fmt.Println(totalItemPriority)
}
